central/declarativeconfig/watch_handler_test.go

```diff
@@ -0,0 +1,301 @@
+			wh := &watchHandler{cachedFileHashes: c.initialCachedFiles}
+			res := wh.compareHashesForChanges(c.fileContents)
+			assert.Equal(t, c.expectedResult, res)
+			assert.Equal(t, c.expectedCachedFiles, wh.cachedFileHashes)
+		})
+	}
+}
+
+func TestWatchHandler_CheckForDeletedFiles(t *testing.T) {
+	cases := map[string]struct {
+		fileContents        map[string][]byte
+		initialCachedFiles  map[string]md5CheckSum
+		expectedResult      bool
+		expectedCachedFiles map[string]md5CheckSum
+	}{
+		"cache containing the file contents should not signal update": {
+			fileContents: map[string][]byte{
+				"test-file": []byte("test content"),
+			},
+			initialCachedFiles: map[string]md5CheckSum{
+				"test-file": md5.Sum([]byte("test content")),
+			},
+			expectedCachedFiles: map[string]md5CheckSum{
+				"test-file": md5.Sum([]byte("test content")),
+			},
+		},
+		"cache containing a deleted file should signal update": {
+			fileContents: map[string][]byte{
+				"test-file": []byte("test content"),
+			},
+			initialCachedFiles: map[string]md5CheckSum{
+				"test-file":        md5.Sum([]byte("test content")),
+				"second-test-file": md5.Sum([]byte("other test content")),
+			},
+			expectedCachedFiles: map[string]md5CheckSum{
+				"test-file": md5.Sum([]byte("test content")),
+			},
+			expectedResult: true,
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			wh := &watchHandler{cachedFileHashes: c.initialCachedFiles}
+			res := wh.checkForDeletedFiles(c.fileContents)
+			assert.Equal(t, c.expectedResult, res)
+			assert.Equal(t, c.expectedCachedFiles, wh.cachedFileHashes)
+		})
+	}
+}
+
+func TestWatchHandler_WithEmptyDirectory(t *testing.T) {
+	// 0. Create a watch handler with a lower interval.
+	updaterMock := mocks.NewMockdeclarativeConfigReconciler(gomock.NewController(t))
+
+	wh := newWatchHandler(updaterMock)
+	opts := k8scfgwatch.Options{
+		Interval: 10 * time.Millisecond,
+	}
+
+	dirToWatch := t.TempDir()
+	// 1. Start the watch handler. Specifically check the returned error, as we do not specify the force flag in the
+	// 	  options.
+	err := k8scfgwatch.WatchConfigMountDir(context.Background(), dirToWatch, k8scfgwatch.DeduplicateWatchErrors(wh), opts)
+	require.NoError(t, err)
+
+	// 2. Add a valid YAML file to the directory the handler is watching.
+	role := declarativeconfig.Role{
+		Name:          "Head Master",
+		Description:   "Head Master of Hogwarts",
+		AccessScope:   "Hogwarts",
+		PermissionSet: "Everything",
+	}
+	yamlBytes, err := yaml.Marshal(&role)
+	require.NoError(t, err)
+
+	// 2.1 Set the expected calls to the updater.
+	updaterMock.EXPECT().ReconcileDeclarativeConfigs([][]byte{yamlBytes})
+
+	filePath := path.Join(dirToWatch, "role")
+	f, err := os.Create(filePath)
+	require.NoError(t, err)
+	defer utils.IgnoreError(f.Close)
+
+	_, err = f.Write(yamlBytes)
+	require.NoError(t, err)
+
+	// 3. Wait to ensure the watch handler has been triggered.
+	time.Sleep(100 * time.Second)
```

Comment:
I'm not sure we want to wait 100 seconds or use `Sleep` at all. Perhaps we can wait on a `cachedFileHashes` change with some timeout?

Reply:
The sleep should be `100 *time.Milisecond`, and I'd rather use it compared to watching for changes. For watching changes, we'd need to acquire a read lock, which could interfere with the watch handler routine and make the test flaky.

Reply:
Hmm, how would acquiring a read lock interfere with the watch handler routine? I thought we put these locks in the first place because we allow for parallel updates.

Reply:
Anyway, in case there are concerns with using `wh.mutex.RLock()` you can use `wh.mutex.TryRLock()`

Reply:
Yeah, I guess what I initially had in mind we'd have to do within tests would be to create a ticker where we would have to continuously check for changes - which would've been a bit too much for me to implement over simply using time.Sleep.
But, I suppose I forgot about `assert.Eventually` - so there isn't really an argument for the `time.Sleep` - I removed those.